### ISSUE TITLE
Idea : Auto lag feature generation and ensamble methods

### ISSUE THREAD

**Issue by BP-Test** [BP-Test/01_fresh_vegetable_competition#7]
Idea 1: **Auto lag-feature Generation** 
According to the feature importance graphs of the [initial models](https://github.com/BP-Test/01_fresh_vegetable_competition/blob/dev/notebook/tutorial_32_yasai_2_predict.ipynb), best lag feature vary over vegetables. 
e.g) Daikon depends on last month's mode price, while Spinach heavily depends on the max temperature of 9 months ago.
On top of namely making lag variables, we'd want our model to further create lag features based on the results of the initial model training in an automatic way. The pros and cons of this approach are as follows.

Pros
1. Can capture unique characteristics of vegetables automatically.
2. Good sample of mlflow experiment control.
3. (Expected to ) Increase the performance of the models .
4. Can reuse same approach to other projects/competitions.
5. Tree methods can be more flexible to timeseries.
6. Can be scaled.

Cons
1. Can be computationally heavy.
2. Will take time to develop
3. Source of bug.
4. Overfitting

**Steps**
1. Run Model with random initialized lags (such as n-1month, n-3 month, n-6 month etc,,)
2. Build a model and evaluate
3. Save top lag features in a dictionary such as {'spinacchi': n-1,n-6,n-9, 'potato': n-6,..} 
5. Create a list of top lags by voting (this could be weighted based on the distribution of samples among vegetables.
6. Create additional lag features.
7. Build model and evaluate
8. Iterate this until metric is no longer improving.

**Requirements**
1. lag_feature_generation function
2. Decorator function @decorator_function

idea2: **Ensemble Method**
XGBoost tend to be computationally heavy when N(number of features) is large, while LightGBM is computationally lighter than XGBoost. We want to maximize the features of these models and improve model.

1. XGBoost : Try various set of lags to come up with best ones
2. LightGBM: Use Auto-lag feature approach on 1.
3. Ensemble 2 models.

Pros :
- Can expect improvements in the model performance
- Simply Interesting : Nice way of good use of the characteristics of the models.
- Can be applied on other models. Such as LSTM + Xgboost etc
- Good usecase of mlflow

Cons:
- Computationally heavy
- Have to write a lot of codes
- Can be buggy
- Overfitting

**Comment by BP-Test** [comment]
@S10aki95  Can you give your thoughts on it??

日本語でいいよ

**Comment by S10aki95** [comment]
@ idea1: 
おけおけ。時間あれば、ちょっと今のコードを一般化する形でやってみる。

@ idea2:
各実装にどれくらい時間かかるかだけど…。できる限りモデルの中身以外はテンプレート化したい。これもちょっと意識しながらやってみる。
